Update Points fundamentals and add an application

kevinyamauchi opened the pull request:
This PR updates the tutorials related to the points layer. In particular, we:

- update the points "fundamentals" to include properties and setting colors with properties.
- add an "application" demonstrating how the point layer can be used to make a simple video annotation app (see below and [this repo](https://github.com/kevinyamauchi/PointAnnotator))

![Mar-23-2020 17-38-11](https://user-images.githubusercontent.com/1120672/77340253-511c8180-6d2d-11ea-927d-dd2d37536605.gif)

This PR requires [#1069 in napari](https://github.com/napari/napari/pull/1069), so we should wait for that to land until merging this PR.

kevinyamauchi commented:
I think this is finally ready for a first review, @napari/core-devs ! I will keep the PR in draft mode because we need to wait for napari 0.3.2 to merge since the application relies on a bugfix that will be released with 0.3.2.

sofroniewn commented on fundamentals/points.md:
```diff
@@ -68,6 +104,9 @@ layer : napari.layers.Points
 
 The input data to the points layer must be an NxD numpy array containing the coordinates of N points in D dimensions. The ordering of these dimensions is the same as the ordering of the dimensions for image layers. This array is always accessible through the `layer.data` property and will grow or shrink as new points are either added or deleted.
 
+## points properties
```
Should this maybe be `Using the points properties dictionary`. It's such amazing functionality, but I worry it's getting lost in the long list of everything else

sofroniewn replied:
I think you should also add something about the points properties dictionary to the top section at the end of this paragraph

> The points layer allows you to display an NxD array of N points in D coordinates. You can adjust the size, face color, edge color of all of the points independently. You can also adjust the opactiy, edge width, and symbol representing all the points simultaneously.

Just to really emphasize this functionality

sofroniewn replied:
I'd also add a link here to your applications note, maybe with a comment like "to see this in action look here" etc

kevinyamauchi replied:
Great points! I have implemented your suggestions.

sofroniewn commented on applications/annotate_points.md:
````diff
@@ -0,0 +1,383 @@
+        
+```
+## point_annotator()
+We will create the GUI within a function called `point_annotator()`. Wrapping the GUI creation in the function allows us to  See below for the function definition.
````
Something has gone wrong with this sentence, end cut off maybe

kevinyamauchi replied:
Good catch. Fixed!

sofroniewn commented on applications/annotate_points.md:
````diff
@@ -0,0 +1,383 @@
+stack = imread(im_path)
+```
+
+We can then start the viewer Note that we use the `napari.gui_qt()` context manager to start and manage Qt event loop. We must provide the `contrast_limits` and set `is_pyramid=False` to prevent the image stack from being evaluated (i.e., loaded) upon layer creation. All of the following GUI code should be within the `napari.gui_qt()` context manager.
````
This kwarg is now `is_multiscale`, but that shouldn't actually be needed anymore and isn't in your function below. Should drop line

kevinyamauchi replied:
Fixed!

sofroniewn commented on applications/annotate_points.md:
````diff
@@ -0,0 +1,383 @@
+We will annotate the features of interest using points in a napari Points layer. Each feature will be given a different label so that we can track them across frames. To achieve this, we will store the label in the `Points.properties` property in the 'label' key. We will instantiate the `Points` layer without any points. Therefore, we will initialize `Points.properties` with the key we will use a ...
+
+```python
+properties = {'label': np.array(labels)}
````
maybe nice to say again here what labels is. Also do you need to call `np.array`? Could someone just pass a list of strings?

kevinyamauchi replied:
> Also do you need to call np.array? Could someone just pass a list of strings?

Ah right, we changed it so they user can pass a list of strings. I will update - definitely nicer for the user to just use a list.

kevinyamauchi replied:
> maybe nice to say again here what labels is.

For sure. Thanks for pointing this out. I missed a couple of other bits that carried over from the "old" way we handling properties. The revised section should be clearer. @sofroniewn, would you mind taking a look and letting me know if this is better?

sofroniewn commented on applications/annotate_points.md:
````diff
@@ -0,0 +1,383 @@
+properties = {'label': np.array(labels)}
+```
+
+We then add the points layer to the viewer using the `viewer.add_points()` method. As discussed above, we will  To visualize the feature each points represent, we set the edge color as a color cycle mapped to the `label` property (`edge_color='label'`). 
````
Looks like something go cut off at the end of this sentence too

kevinyamauchi replied:
Fixed!

sofroniewn commented on applications/annotate_points.md:
````diff
@@ -0,0 +1,383 @@
+```
+
+## Adding a UI for selecting points
+Next, we will use magicgui to add a dropdown menu for selecting which point we would like to add. [magicgui](https://github.com/napari/magicgui) is a library for building GUIs from functions and works by applying function decorators. To make the a dropdown menu populated with the valid point labels, we simply define the function with the label as an input argument and then decorate it with the `m ...
````
maybe say `[magicgui](https://github.com/napari/magicgui) is a library for building GUIs from functions from the napari team`. I think it matters that it is made by @tlambert03 and us in terms of signaling to users that it is something they can trust and that we will continue to support

kevinyamauchi replied:
Sounds good!

kevinyamauchi replied:
Fixed!

sofroniewn commented on applications/annotate_points.md:
````diff
@@ -0,0 +1,383 @@
+points_layer.mouse_drag_callbacks.append(next_on_click)
+```
+
+## Saving annotation keybinding
````
I think this can now be done with our builtin csv writer on the points layer? If not would be good to know why and maybe adapt our builtin writer so it can meet the needs of this example. I'd much rather point people towards using the built in then having to add their own `Control-S` keybinding, which will overwrite ours.

sofroniewn replied:
I guess this saving maybe is done with some special deeplabcut multindexing, which makes sense that it is custom, so maybe I'm asking too much for one example, but curious what you make of the idea.

Also curious if it is possible to add hookimpls in scripts like this? Maybe @tlambert03 knows? That would be pretty cool, because then you could add a custom `napari_write_points` hookimpl rather than having to add a custom `Control-S` keybinding

kevinyamauchi replied:
We can definitely write the points using the builtin writer. The function here writes out the DLC format, which I think is a little different from the builtin writer format. I was thinking it would be cool to make a follow up application turning that function into a writer plugin. However, I am also cool with making this tutorial using the built-in and the a follow up one on making a writer plugin.

In either case, you make a good point about the usage of `Control-S`.

sofroniewn replied:
Yeah, maybe let's have a dedicated application note making a DLC reader/ writer plugin - could even think then about contributing it to DLC. For now for this example let's just use the builtin

kevinyamauchi replied:
Sounds good. I think using the built-in writer also helps keep the focus on using the core napari features.

sofroniewn replied:
I think we can actually drop the whole concept of `output_path` from this example and just instruct people to do `Control-S` with the layer selected type in the path. We should have a section explaining this, but I don't think we need to have a custom keybinding. It might be nice to include a screenshot with the save dialog open overlaid on the viewer in this section too

kevinyamauchi replied:
Done!

kevinyamauchi commented:
Thanks for the reviews, @sofroniewn ! I will address your points tomorrow morning!

kevinyamauchi commented:
@sofroniewn , I think I have addressed your comments, it would be great if you could take another look when you have time.

Also, I don't know how to build the pages locally to make sure everything looks right. @kne42  or @tlambert03, would you mind telling me how or pointing me to some instructions?

kevinyamauchi commented:
@sofroniewn , I removed the custom save function and instead provided instructions for using the builtin save dialog.

jni commented on applications/annotate_points.md:
````diff
@@ -0,0 +1,314 @@
+```
+that looks like this:
+
+![image]({{ '/assets/tutorials/point_annotator_demo.gif' | relative_url }})
````
fwiw, it would be nice if everyone got on board with video tags instead of gifs. They use up less space (so download faster), *and* look smoother. An example from my blog post:

```html
<video width="90%" autoplay loop muted playsinline>
  <source src="/napari/napari-blobs-720p.mp4" type="video/mp4" />
  <source src="/napari/napari-blobs-720p.ogg" type="video/ogg" />
  <img src="/napari/napari-blobs-720p-frame.jpg"
      title="Your browser does not support the video tag." />
</video>
```

As a reminder, markdown is a superset of html, so this works fine within a markdown document.

sofroniewn replied:
for video tags does that mean we make `.mp4` instead? say using a tool like the quicktime screen recorder instead of `gifs`? My impression was they were always much larger, but I'm open to suggestions/ ideas here

kevinyamauchi replied:
The video files also end up being larger for me. Perhaps we can circle back to this in the future.

jni commented on applications/annotate_points.md:
```diff
@@ -0,0 +1,314 @@
+    stack = imread(im_path)
+    
+    with napari.gui_qt():
+        viewer = napari.view_image(stack, contrast_limits=[0, 256])
```
is stack uint8? If so, you shouldn't need the `contrast_limits` anymore?

kevinyamauchi replied:
Good point. Fixed

jni commented on applications/annotate_points.md:
```diff
@@ -0,0 +1,314 @@
+            ind = list(labels).index(current_label)
+            new_ind = (ind + 1) % len(labels)
+            new_label = labels[new_ind]
+            current_properties['label'] = np.array([new_label])
```
Hmmm, why is this a list/array rather than an item?

jni commented on applications/annotate_points.md:
```diff
@@ -0,0 +1,314 @@
+        )
+        points_layer.edge_color_mode = 'cycle'
+
+        # Create the label selection menu
+        @magicgui(label={'choices': labels})
+        def label_selection(label):
+            return label
+        label_menu = label_selection.Gui()
+
+        def update_label_menu(event):
+            """Update the label menu when the point selection changes"""
+            label_menu.label = points_layer.current_properties['label'][0]
+
+        points_layer.events.current_properties.connect(update_label_menu)
+
+        def label_changed(result):
+            """Update the Points layer when the label menu selection changes"""
+            selected_label = result
+            current_properties = points_layer.current_properties
+            current_properties['label'] = np.asarray([selected_label])
+            points_layer.current_properties = current_properties
+
+        label_menu.label_changed.connect(label_changed)
```
I'd pull all of this out into its own function, `create_label_menu`, that takes a points layer as an input. I think the amount of code inside the `with napari.gui_qt()` block is huge, so it'd be good to reduce it into manageable chunks.

kevinyamauchi replied:
Good idea! Done.

jni commented on applications/annotate_points.md:
```diff
@@ -0,0 +1,314 @@
+        viewer.window.add_dock_widget(label_menu)
+
+        @viewer.bind_key('.')
+        def next_label(event=None):
```
Similarly, you can actually use the decorator as a function directly, and take points_layer as an input:

```python
def next_label(event=None, layer=None):
    current_properties = layer.current_properties
    ...

with napari.gui_qt():
    viewer = napari.Viewer()
    ...
    viewer.bind_key('.', lambda event: next_label(event, points_layer))
```

kevinyamauchi replied:
I think generally speaking, this would be a nice way to do this. My concern is that the usage of lambda bumps us the "prerequisites" for understanding this tutorial. If possible, I would like to keep the level of entry as low as possible for this tutorial (and most napari tutorials).

I think we will soon need to write a series of more advanced tutorials on building an application with/around napari (e.g., extending the GUI, keybindings, plugins). Perhaps we can include tips like this in one of those.

jni commented on applications/annotate_points.md:
```diff
@@ -0,0 +1,314 @@
+            current_properties['label'] = np.array([new_label])
+            points_layer.current_properties = current_properties
+
+        def next_on_click(layer, event):
```
This doesn't need to be nested at all.

jni commented on applications/annotate_points.md:
```diff
@@ -0,0 +1,314 @@
+
+
+        @viewer.bind_key(',')
+        def prev_label(event):
```
As above, this can be unnested by taking in points_layer as a parameter and using a lambda.

jni commented on applications/annotate_points.md:
````diff
@@ -0,0 +1,314 @@
+```
+
+## Annotating with points
+We will annotate the features of interest using points in a napari Points layer. Each feature will be given a different label so that we can track them across frames. To achieve this, we will store the label in the `Points.properties` property in the 'label' key. We will instantiate the `Points` layer without any points. However, we will initialize `Points.properties` with the property values we  ...
````
the `Points.properties` property -> the `Points.properties` attribute?

also, "we will annotating" -> "we will be annotating"

kevinyamauchi replied:
Fixed!

jni commented on applications/annotate_points.md:
````diff
@@ -0,0 +1,314 @@
+```python
+properties = {'label': labels}
+points_layer = viewer.add_points(
+    properties=properties,
````
As mentioned in the review of the complete code, I think you could just inline `{'label': labels}`

kevinyamauchi replied:
Fixed!

jni commented on applications/annotate_points.md:
````diff
@@ -0,0 +1,314 @@
+To create the GUI, we call the label_selection. Gui() method.
+
+```python
+label_menu = label_selection.Gui()
````
btw @tlambert03 why the choice `Gui` instead of `gui`?

tlambert03 replied:
it's a class... so I went with caps.

jni commented on applications/annotate_points.md:
```diff
@@ -0,0 +1,314 @@
+### Saving the annotations
+Once we are happy with the annotations, we can save them to a CSV file using the builing CSV writer for the points layer. To do so, first, select the "Points" layer in the layer list and then click "Save Selected layer(s)"  in the "File" menu or press control+S (cmd+S on Mac OS)  to bring up the file save dialog. From here you can enter the file path and save the annotation coordinates as a CSV.
+
+![image]({{ '/assets/tutorials/points_save_dialog.png' | relative_url }})
```
We can also use `points_layer.save('path/to/file.csv')`, which I think is worth pointing out!

kevinyamauchi replied:
I can add some text mentioning the `points_layer.save()` method.

jni commented on fundamentals/points.md:
```diff
@@ -113,6 +154,88 @@ Individual points can each have different edge and face colors. You can initiall
+* PuGr
+* GrBu
+* GrBu_d
+* RdBu
```
Is this list accurate/exhaustive? Where are magma, inferno, plasma? Where is twilight? These certainly don't correspond to our image colormaps...

sofroniewn replied:
They were used here https://napari.org/tutorials/fundamentals/image, we should probably correct them in both places or remove the full long list from these pages and link to a new page dedicated around our colormaps

kevinyamauchi replied:
Yeah, that's where I pulled the list from. I can leave them as is and then we can make a separate PR with a page for colormaps.

kevinyamauchi replied:
I made an issue here: https://github.com/napari/tutorials/issues/71

kevinyamauchi commented:
As always, thank you for your thorough review, @jni ! At first glance the suggestions seem reasonable and not onerous at all. I will take address them before merging.

sofroniewn commented:
@kevinyamauchi can you finish this one up, because I'd like to reference it in my neubias talk next week, i think it's pretty close!

kevinyamauchi commented:
@sofroniewn , no problem! It's on the list for today. You should have it by morning your time.

kevinyamauchi commented:
@sofroniewn , I think this is good to go! I will let it sit for a bit and merge tonight.
